Add restrictions for preventing undesirable deploys

sihil opened the pull request:
On of the features of Frontend's `goo` wrapper around riff-raff is the ability to block deploys. In addition @adamnfish has expressed a desire to lock down the ability to deploy Janus. 

This goes someway to achieve both of these and hopefully provide just enough flexibility that it will be useful more widely for temporarily blocking deploys.

Further features could be added (or moved to issues for future consideration):
- Nothing currently stops someone creating a new continuous deploy for a restricted project that permits continuous deployments
- Link from a deploy to create a restriction
- Indication of restriction as soon as you enter the project on the deploy screen

adamnfish commented on riff-raff/app/deployment/Deployments.scala:
```diff
@@ -21,25 +22,29 @@ import scala.util.control.NonFatal
+      requestedParams.build.projectName, requestedParams.stage.name, requestSource)
+    if (restrictionsPreventingDeploy.nonEmpty) {
+      Left(s"Unable to queue deploy as restrictions are currently in place: ${restrictionsPreventingDeploy.map(r => s"${r.fullName}: ${r.note}").mkString("; ")}")
+    } else if (enableQueueingSwitch.isSwitchedOff)
```
:minor_point: Why omit the braces from this `else if` clause?

sihil replied:
Good spot.

adamnfish commented on riff-raff/app/deployment/model.scala:
```diff
@@ -3,11 +3,19 @@ package deployment
 import org.joda.time.format.PeriodFormatterBuilder
 import utils.VCSInfo
 
+trait RequestSource
```
Should this be sealed?

adamnfish commented on riff-raff/app/restrictions/RestrictionChecker.scala:
```diff
@@ -0,0 +1,49 @@
+    * @tparam T The function return type
+    * @return The result of either ifEditable or ifNotEditable
+    */
+  def editable[T](currentConfig: Option[RestrictionConfig], identity: UserIdentity)
```
Does this really save typing compared to returning a Boolean and using an if statement? If statements also take two blocks - the one to run if it is true and the one to run if it is not.

sihil replied:
Removed.

adamnfish commented on riff-raff/app/restrictions/RestrictionConfig.scala:
```diff
@@ -0,0 +1,32 @@
+  * A configuration that restricts which deploys are allowed
+  * @param id Unique ID for CRUD storage and retrival
+  * @param projectName The project name that this applies to
+  * @param stage The stage that this applies to (this can be a regex, but if no special regex characters are found then
```
Would it be better to have this be a regex type field? A string literal is a subset of a the regex type and this would move the "if no special character" logic to the edge.

sihil replied:
Regex doesn't support Scala like equality (i.e. "a".r != "a".r) so I tend not to use it as a field in case classes  - it breaks the case class semantics.

adamnfish commented on riff-raff/app/views/restrictions/form.scala.html:
```diff
@@ -0,0 +1,81 @@
+    <div class="row">
+        <div class="col-md-7">
+            @b3.form(action=routes.Restrictions.save) {
+            @CSRF.formField
```
:-)

sihil replied:
😄  - `b3` actually has a `formCSRF` that does this for you so I've just switched over.

philwills commented on riff-raff/app/ci/ContinuousDeployment.scala:
```diff
@@ -48,7 +48,10 @@ class ContinuousDeployment(deployments: Deployments) extends Lifecycle with Logg
         log.info(s"Triggering deploy of ${params.toString}")
         try {
-          deployments.deploy(params)
+          deployments.deploy(params, requestSource = ContinuousDeploymentRequestSource) match {
```
Would a `leftMap` (https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/syntax/either.scala#L120) be more readable here, or would that make you think we cared about the result value?

sihil replied:
Ah yes - exactly what I did after @philwills similar comment.

philwills commented on riff-raff/app/controllers/DeployController.scala:
```diff
@@ -73,7 +73,10 @@ class DeployController(deployments: Deployments, prismLookup: PrismLookup, deplo
-            val uuid = deployments.deploy(parameters)
+            val uuid = deployments.deploy(parameters, requestSource = UserRequestSource(request.user)) match {
+              case Right(uuid) => uuid
+              case Left(error) => throw new IllegalStateException(error)
```
I'm assuming this is a case of minimal viable interface.

sihil replied:
Ideally we'd eliminate the throwing of exceptions entirely, but pulling the exception throwing up to the top level is a step in the right direction without making too many sweeping changes.

I've tweaked this to use `valueOr` from the cats either syntax.

adamnfish commented on riff-raff/app/views/restrictions/list.scala.html:
```diff
@@ -0,0 +1,62 @@
+                    <th>Stage</th>
+                    <th>Locked</th>
+                    <th></th>
+                    <th></th>
```
If this `<th>` doesn't mean anything by itself you could colspan it with the previous one as the header for "actions".

:minor_point:

adamnfish commented on riff-raff/test/restrictions/RestrictionCheckerTest.scala:
```diff
@@ -0,0 +1,168 @@
+import org.joda.time.DateTime
+import org.scalatest.{FlatSpec, Matchers}
+
+class RestrictionCheckerTest extends FlatSpec with Matchers {
```
There are a lot of tests here for "does my version of an if statement work like an if statement"

sihil replied:
With the change of signature and use these are now saner tests.

philwills commented on riff-raff/app/controllers/api.scala:
```diff
@@ -294,21 +293,30 @@ class Api(deployments: Deployments, deploymentTypes: Seq[DeploymentType])(implic
+                "logURL" -> routes.DeployController.viewUUID(deployId.toString).absoluteURL()
+              )
+            )
+          case Left(error) =>
```
This feels like the job of some transformer from an API result type rather than the responsibility of a particular endpoint.

sihil replied:
Agreed - `ApiJsonEndpoint` should probably take something like `Request => Either[ApiError, ApiResult]`.

Not in this PR though, do you want to raise an issue?

philwills commented on riff-raff/app/deployment/Deployments.scala:
```diff
@@ -21,25 +22,29 @@ import scala.util.control.NonFatal
+    val restrictionsPreventingDeploy = RestrictionChecker.configsThatPreventDeployment(RestrictionConfigDynamoRepository,
+      requestedParams.build.projectName, requestedParams.stage.name, requestSource)
+    if (restrictionsPreventingDeploy.nonEmpty) {
+      Left(s"Unable to queue deploy as restrictions are currently in place: ${restrictionsPreventingDeploy.map(r => s"${r.fullName}: ${r.note}").mkString("; ")}")
```
I'd rather we had a type other than `String` for this, but happy for that to come later.

sihil replied:
It was easy to do so I've created a small `case class Error(message: String) extends AnyVal`.

philwills commented on riff-raff/app/persistence/RestrictionConfigDynamoRepository.scala:
```diff
@@ -0,0 +1,22 @@
+import com.gu.scanamo.Table
+import restrictions.{RestrictionConfig, RestrictionsConfigRepository}
+
+object RestrictionConfigDynamoRepository extends DynamoRepository with RestrictionsConfigRepository {
```
I hate this kind of redundant interface, but the alternative isn't as simple as it should be without lots of wider changes.

sihil replied:
Testing is non-trivial without it. Having a local dynamodb for tests would also eliminate it.

philwills replied:
I'll look to add https://github.com/localytics/sbt-dynamodb

philwills commented on riff-raff/app/restrictions/RestrictionChecker.scala:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  def stageMatches(configStage: String, stage: String): Boolean =
+    if (configStage.matches(""".*[$^.+*?()\[{|].*""")) {
```
Yikes!

adamnfish commented on riff-raff/app/controllers/Restrictions.scala:
```diff
@@ -0,0 +1,86 @@
+      val canSave = RestrictionChecker.editable(Some(rc), request.user)(true){_=>false}
+      Ok(views.html.restrictions.form(restrictionsForm.fill(
+        RestrictionForm(rc.id, rc.projectName, rc.stage, rc.editingLocked, rc.whitelist, rc.continuousDeployment, rc.note)
+      ), saveDisabled = !canSave))
```
This is quite an important piece of logic and it's hidden at the back of the arg list behind a `!`.

adamnfish replied:
(see also above)

sihil replied:
Fairly important, although this is for display purposes only rather than enforcing it. I've rearranged the code so it's clearer.

adamnfish commented on riff-raff/app/controllers/Restrictions.scala:
```diff
@@ -0,0 +1,86 @@
+  }
+  def edit(id: String) = AuthAction { implicit request =>
+    RestrictionConfigDynamoRepository.getRestriction(UUID.fromString(id)).map{ rc =>
+      val canSave = RestrictionChecker.editable(Some(rc), request.user)(true){_=>false}
```
You should assign the boolean in the state you mean to use it. `cannotSave`, in this case.

adamnfish commented:
@philwills yeah, I mentioned IRL having admins that can override a locked deployment might be handy. That said, I think in the security case of "I intend for the only way to deploy this to be continuous deployment" it makes sense to be able to lock it down to some extent.

TBonnin commented:
As discussed in person yesterday, it seems to me the use case for blocking is more at deploy level rather than project level, but I understand Riff-Raff doesn't keep track of deploy entities at the moment.
For example it mean that blocking multi-projects deploy like Frontend would requires creating multiple restrictions. Not ideal but fine for now I guess.

sihil commented on riff-raff/app/controllers/DeployController.scala:
```diff
@@ -73,7 +73,10 @@ class DeployController(deployments: Deployments, prismLookup: PrismLookup, deplo
             )
           case "deploy" =>
-            val uuid = deployments.deploy(parameters)
+            val uuid = deployments.deploy(parameters, requestSource = UserRequestSource(request.user)) match {
```
I've used cats' `valueOr` pimp from `cats.syntax.either._`.

sihil commented on riff-raff/app/controllers/Restrictions.scala:
```diff
@@ -0,0 +1,86 @@
+        f.continuousDeployment,  f.note))
+    ).verifying(
+      "Stage is invalid - should be a valid regular expression or contain no special values",
+      form => try { form.stage.r; true } catch { case e:PatternSyntaxException => false }
```
I thought that would catch fatal errors - but actually Try has that covered which I hadn't realised. Nice idea.

sihil commented on riff-raff/app/controllers/Restrictions.scala:
```diff
@@ -0,0 +1,86 @@
+    Ok(views.html.restrictions.form(newForm, saveDisabled = false))
+  }
+
+  val editableOrForbidden: (Option[RestrictionConfig], UserIdentity) => (=> Result) => Result =
```
I **_am_** trying to be clever.

sihil commented on riff-raff/app/controllers/Restrictions.scala:
```diff
@@ -0,0 +1,86 @@
+    restrictionsForm.bindFromRequest().fold(
+      formWithErrors => Ok(views.html.restrictions.form(formWithErrors, saveDisabled = false)),
+      f => {
+        editableOrForbidden(RestrictionConfigDynamoRepository.getRestriction(f.id), request.user) {
```
Fair comment. I've made various modifications to address this.

sihil commented:
@TBonnin I think that's the right level for a restriction usually. I'm hoping that projects and deploys should match much more closely once teams move to riff-raff.yaml. Potentially adding API endpoints for adding and removing restrictions might be helpful for doing batches in the case where this is not enough - or further improvements (potentially regexes on the project name too?)

sihil commented:
OK - I've addressed all of the feedback and I've also added a really simple superuser feature that allows users in a list of e-mail addresses (in the config) to edit any configuration. We can add half a dozen or so users to this list as a safeguard against people abusing the locking feature.
